launch/launch/utilities/type_utils.py

```diff
@@ -0,0 +1,451 @@
+# Copyright 2019-2020 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Module which implements get_typed_value function."""
+
+import collections.abc
+from typing import Any
+from typing import cast
+from typing import List
+from typing import Optional
+from typing import Sequence
+from typing import Set
+from typing import Text
+from typing import Tuple
+from typing import Type
+from typing import Union
+
+import yaml
+
+from .normalize_to_list_of_substitutions_impl import normalize_to_list_of_substitutions
+from .perform_substitutions_impl import perform_substitutions
+from ..launch_context import LaunchContext
+from ..some_substitutions_type import SomeSubstitutionsType
+from ..substitution import Substitution
+
+ScalarTypesTuple = (
+    int, float, bool, str
+)
+
+ListTypesTuple = (
+    List[int], List[float], List[bool], List[str]
+)
+
+AllowedTypesTuple = ScalarTypesTuple + ListTypesTuple
+
+"""Allowed scalar types."""
+ScalarTypesType = Type[Union[int, float, bool, str]]
+"""Allowed uniform list types."""
+ListTypesType = Type[Union[
+    List[int], List[float], List[bool], List[str]
+]]
+"""Allowed types."""
+AllowedTypesType = Union[ScalarTypesType, ListTypesType]
```

Comment:
> I understand why the focus on uniform lists, but this is clearly at odds with Parser.get_attr() claiming it can handle non-uniform lists.

Non uniform lists add extra complexity, and I can't think of an use case for them.
I have to update old doc blocks.

Reply:
> Hmm, a mixed integer/float

The coercion functionality is flexible, and a substitution without a trailing `.` can be coerced to a `float`:
https://github.com/ros2/launch/blob/7cc161161eaff98b1f6cd3068148af9ed29fea00/launch/test/launch/utilities/test_type_utils.py#L133
But the checking functionality is more strict: https://github.com/ros2/launch/blob/7cc161161eaff98b1f6cd3068148af9ed29fea00/launch/test/launch/utilities/test_type_utils.py#L133

Python behaves in a similar way:
```python3
assert isinstance(3, float) == False
assert float('3') == 3.
```

If we want, we can make the system more flexible in the future.